integration/e2e/metrics_promql.go

```diff
@@ -0,0 +1,283 @@
+	case promql.Scalar:
+		return s.V, nil
+
+	case promql.Vector:
+		if len(s) == 0 {
+			return 0, nil
+		}
+
+		if len(s) == 1 {
+			return s[0].V, nil
+		}
+
+		return 0, fmt.Errorf(errTooManyVectorElements, len(s))
+	}
+
+	return 0, errNotScalarOrVector
+}
+
+type sample struct {
+	labels map[string]string
+	ts     int64
+	val    float64
+}
+
+func convertMetricFamiliesToSamples(ts int64, mfs map[string]*dto.MetricFamily) []sample {
+	result := []sample{}
+
+	for _, mf := range mfs {
+		switch mf.GetType() {
+		case dto.MetricType_COUNTER:
+			for _, m := range mf.GetMetric() {
+				result = append(result, sample{
+					labels: getLabels(mf.GetName(), m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetCounter().GetValue(),
+				})
+			}
+
+		case dto.MetricType_GAUGE:
+			for _, m := range mf.GetMetric() {
+				result = append(result, sample{
+					labels: getLabels(mf.GetName(), m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetGauge().GetValue(),
+				})
+			}
+
+		case dto.MetricType_SUMMARY:
+			for _, m := range mf.GetMetric() {
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_count", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    float64(m.GetSummary().GetSampleCount()),
+				})
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_sum", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetSummary().GetSampleSum(),
+				})
+			}
+
+		case dto.MetricType_HISTOGRAM:
+			for _, m := range mf.GetMetric() {
+				for _, b := range m.GetHistogram().GetBucket() {
+					result = append(result, sample{
+						labels: getLabelsExt(mf.GetName()+"_bucket", m.GetLabel(), true, b.GetUpperBound()),
+						ts:     getTimestamp(ts, m.GetTimestampMs()),
+						val:    float64(b.GetCumulativeCount()),
+					})
+				}
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_count", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    float64(m.GetHistogram().GetSampleCount()),
+				})
+
+				result = append(result, sample{
+					labels: getLabels(mf.GetName()+"_sum", m.GetLabel()),
+					ts:     getTimestamp(ts, m.GetTimestampMs()),
+					val:    m.GetHistogram().GetSampleSum(),
+				})
+			}
+		}
+	}
+
+	return result
+}
+
+func getTimestamp(ts int64, mts int64) int64 {
+	if mts != 0 {
+		return mts
+	}
+	return ts
+}
+
+func getLabels(name string, pairs []*dto.LabelPair) map[string]string {
+	return getLabelsExt(name, pairs, false, 0)
+}
+
+func getLabelsExt(name string, pairs []*dto.LabelPair, bucket bool, upperBound float64) map[string]string {
+	result := map[string]string{}
+	result[labels.MetricName] = name
+	if bucket {
+		result[labels.BucketLabel] = fmt.Sprintf("%g", upperBound)
+	}
+
+	for _, p := range pairs {
+		result[p.GetName()] = p.GetValue()
+	}
+	return result
+}
+
+// implements Queryable and Querier
```

Comment:
Instead of this comment (which is not proper full sentence), maybe it's better document by code?

```
var (
    _ Queryable = samplesQueryable{}
    _ Querier = samplesQueryable{}
)
```